src/EditorFeatures/Core/InheritanceMargin/InheritanceMarginServiceHelpers.cs

```diff
@@ -138,20 +145,36 @@ internal static class InheritanceMarginServiceHelper
             // Go down the inheritance chain to find all the overrides targets.
             var allOverriddenSymbols = await SymbolFinder.FindOverridesArrayAsync(memberSymbol, solution, cancellationToken: cancellationToken).ConfigureAwait(false);
 
-            // Go up the inheritance chain to find all the implemented targets.
-            var implementingSymbols = memberSymbol.ExplicitOrImplicitInterfaceImplementations();
-
             // Go up the inheritance chain to find all overriding targets
             var overridingSymbols = GetOverridingSymbols(memberSymbol);
 
+            // Go up the inheritance chain to find all the implemented targets.
+            var allImplementingSymbols = GetImplementingSymbolsForTypeMember(memberSymbol, overridingSymbols);
+
             // Go down the inheritance chain to find all the implementing targets.
-            var allImplementedSymbols = await GetImplementedSymbolsAsync(solution, memberSymbol, cancellationToken).ConfigureAwait(false);
+            var allImplementedSymbols = await GetImplementedSymbolsForTypeMemberAsync(solution, memberSymbol, cancellationToken).ConfigureAwait(false);
 
             // For all overriden & implemented symbols, make sure it is in source.
             // For example, if the user is viewing System.Threading.SynchronizationContext from metadata,
             // then don't show the derived overriden & implemented method in the default implementation for System.Threading.SynchronizationContext in metadata
             var overriddenSymbols = allOverriddenSymbols.WhereAsArray(symbol => symbol.Locations.Any(l => l.IsInSource));
-            var implementedSymbols = allImplementedSymbols.WhereAsArray(symbol => symbol.Locations.Any(l => l.IsInSource));
+
+            // Make sure all implemented symbols are unique.
+            // For example:
+            // interface IBar { void Foo(); }
+            // class Bar : IBar { public virtual void Foo() { } }
+            // class Bar2 : Bar, IBar { public override void Foo() { } }
+            // Here Bar2.Foo() would be added twice as an implemented members.
+            var implementedSymbols = allImplementedSymbols.Distinct().WhereAsArray(symbol => symbol.Locations.Any(l => l.IsInSource));
+
+            // Make sure the symbols are unique for implementing symbols
+            // For example:
+            // interface IBar { void Foo(); }
+            // class Bar : IBar { public virtual void Foo() { } }
+            // class Bar2 : Bar, IBar { public override void Foo() { } }
+            // Here IBar.Foo() would be added twice as an implementing members.
+            // The order is not important, because UI would sort it based on display name.
+            var implementingSymbols = allImplementingSymbols.SelectAsArray(symbol => symbol.OriginalDefinition).Distinct();
```

Comment:
so my recommendation would be to avoid all of the Select/Distinct *here*.  Instead, just move into CreateInheritanceMemberInfoForMemberAsync where it just applies one helper method to all 4 arrays it gets passed in.

or, if you want to do this here.  have the select/distinct happen just prior to this 'if' check.  if you want me to make changes that i'm talking about, i can show you

Reply:
I think this time everything should be on the right track now. : )